builtin/v13/migration/market.go

```diff
@@ -0,0 +1,548 @@
+		}
+		actorSectorsMapRoot = typegen.CborCid(newActorSectorsMapRoot)
+
+		if err := prevOutProviderSectors.Put(abi.UIntKey(uint64(miner)), &actorSectorsMapRoot); err != nil {
+			return cid.Undef, cid.Undef, xerrors.Errorf("failed to put actor sectors: %w", err)
+		}
+	}
+
+	// get roots
+	outProviderSectors, err := prevOutProviderSectors.Root()
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to get providerSectorsRoot: %w", err)
+	}
+
+	outStates, err := prevOutStates.Root()
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to get statesRoot: %w", err)
+	}
+
+	return outProviderSectors, outStates, nil
+}
+
+func (m *marketMigrator) migrateProviderSectorsAndStatesFromScratch(ctx context.Context, store cbor.IpldStore, in migration.ActorMigrationInput, states cid.Cid) (cid.Cid, cid.Cid, error) {
+	ctxStore := adt.WrapStore(ctx, store)
+
+	oldStateArray, err := adt.AsArray(ctxStore, states, market12.StatesAmtBitwidth)
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load states array: %w", err)
+	}
+
+	newStateArray, err := adt13.MakeEmptyArray(ctxStore, market13.StatesAmtBitwidth)
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load states array: %w", err)
+	}
+
+	providerSectorsMem := map[abi.ActorID]map[abi.SectorNumber][]abi.DealID{}
+
+	var oldState market12.DealState
+	var newState market13.DealState
+
+	// FIP: For each deal state object in the market actor state that has a terminated epoch set to -1:
+	err = oldStateArray.ForEach(&oldState, func(i int64) error {
+		deal := abi.DealID(i)
+
+		newState.SlashEpoch = oldState.SlashEpoch
+		newState.LastUpdatedEpoch = oldState.LastUpdatedEpoch
+		newState.SectorStartEpoch = oldState.SectorStartEpoch
+		newState.SectorNumber = 0 // non- -1 slashed epoch
+
+		if oldState.SlashEpoch == -1 {
+			// FIP: find the corresponding deal proposal object and extract the provider's actor ID;
+			// - we do this by collecting all dealIDs in providerSectors in miner migration
+
+			// in the provider's miner state, find the ID of the sector with the corresponding deal ID in sector metadata;
+			sid, ok := m.providerSectors.dealToSector[deal]
+			if ok {
+				newState.SectorNumber = sid.Number // FIP: set the new deal state object's sector number to the sector ID found;
+
+				// FIP: add the deal ID to the ProviderSectors mapping for the provider's actor ID and sector number.
+				if _, ok := providerSectorsMem[sid.Miner]; !ok {
+					providerSectorsMem[sid.Miner] = make(map[abi.SectorNumber][]abi.DealID)
+				}
+				providerSectorsMem[sid.Miner][sid.Number] = append(providerSectorsMem[sid.Miner][sid.Number], deal)
+			} else {
+				//fmt.Printf("deal %d not found in providerSectors: %v\n", deal, oldState)
+
+				newState.SectorNumber = 0 // FIP: if such a sector cannot be found, assert that the deal's end epoch has passed and use sector ID 0
```

Comment:
This isn't making the assertion that the FIP specifies. If there's a good reason for that, can it be written in the comment? Should the FIP be changed?

Reply:
I believe the assert is caught by this invariant

```
acc.Require((deal.SectorNumber == sectorDeal.SectorNumber) || (deal.SectorNumber == 0 && deal.SlashEpoch != -1),
			"deal sector number %d does not match sector %d for miner %v (ds: %#v; ss %#v)",
```

Will add a comment here